### PR TITLE
Fix invariant propagation

### PR DESCRIPTION
**Pull request by mabaro**
GLSL/SPIRV output seems to work fine (i.e., no problems/glitches detected when using _invariant_), thus only for MSL output invariant position dependencies are forced into temporaries.

### CODE REVIEW

**Comment by mabaro** [comment]
@HansKristian-Work These bits were missing from my previous pull request in order to make it work for MSL

File: spirv_glsl.cpp [hunk 1]
```diff
@@ -11766,8 +11766,8 @@ void CompilerGLSL::disallow_forwarding_in_expression_chain(const SPIRExpression
 	// Allow trivially forwarded expressions like OpLoad or trivial shuffles,
 	// these will be marked as having suppressed usage tracking.
 	// Our only concern is to make sure arithmetic operations are done in similar ways.
-	if (expression_is_forwarded(expr.self) && !expression_suppresses_usage_tracking(expr.self) &&
-	    forced_invariant_temporaries.count(expr.self) == 0)
+	if( ( forced_invariant_temporaries.count( expr.self ) == 0 ) &&
+		( backend.force_invariant_temporaries || ( expression_is_forwarded( expr.self ) && !expression_suppresses_usage_tracking( expr.self ) ) ) )
```
Comment (HansKristian-Work) [inline spirv_glsl.cpp:11770]: Can you explain what the real difference is here? The only difference I can see is that the previous code has an escape hatch for expression_suppresses_usage_tracking.

Is the fix here that for MSL, we should ignore that?

**Comment by HansKristian-Work** [comment]
Please make sure to sign the CLA, otherwise I cannot merge anything.

File: spirv_glsl.cpp (hunk 1 above)
Comment (mabaro) [reply spirv_glsl.cpp:11770]: Without these **forced_invariant_temporaries** the invariant propagation  (i.e., forcing temporaries) stops before reaching the end of the dependency chain.
The intent of this option is to force the whole dependency chain into temporaries, only for MSL.
GLSL/SPIRV output (i.e, android) has not shown this kind of problems so far, thus I'd leave that codepath unchanged unless you think that would also make sense.

Comment (HansKristian-Work) [reply spirv_glsl.cpp:11770]: Oh, I see now.

Comment (HansKristian-Work) [reply spirv_glsl.cpp:11770]: Does https://github.com/KhronosGroup/SPIRV-Cross/pull/2428 resolve the issue?

Comment (mabaro) [reply spirv_glsl.cpp:11770]: wrt #2428:
On a first check on a single shader abusing CompilerGLSL::flags_to_qualifiers_glsl() to output _INV_ tag it propagates nicely until the beginning of main().
Full rebuild seems good as well, no glitches.

**Comment by HansKristian-Work** [comment]
Merged as part of #2428.

**Pull request closed without merging**